docs: add section to upgrade guide on lazy loading AngularJS

brandonroberts opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

brandonroberts committed:
docs: add section to upgrade guide on lazy loading AngularJS

mary-poppins commented:
You can preview c266e71 at https://pr30541-c266e71.ngbuilds.io/.

gkalpak commented on aio/content/examples/upgrade-lazy-load-ajs/src/app/angularjs-app/index.ts:
```diff
@@ -0,0 +1,31 @@
+import * as angular from 'angular';
+import 'angular-route';
+
+import 'jquery';
```
Why use jQuery?

gkalpak commented on aio/content/examples/upgrade-lazy-load-ajs/src/app/app.component.ts:
```diff
@@ -0,0 +1,10 @@
+  styleUrls: ['./app.component.css']
+})
+export class AppComponent {
+  title = 'angularjs-hybrid';
```
Is this used anywhere?

gkalpak commented on aio/content/examples/upgrade-lazy-load-ajs/src/app/lazy-loader.service.ts:
```diff
@@ -0,0 +1,19 @@
+  providedIn: 'root'
+})
+export class LazyLoaderService {
+  bootstrapped = false;
```
Is this used anywhere?
Did you mean to use it to avoid bootstrapping the AngularJS app multiple times? (You probably should, because bootstrapping it multiple times will probably break :grin:)

brandonroberts replied:
Yep 😄

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -838,6 +838,59 @@ After this, the service is injectable anywhere in AngularJS code:
+
+When migrating large applications from AngularJS to Angular using a hybrid approach, you want to migrate some of the most commonly used features first. For example, a YouTube client application might migrate the home page, then the player page, then search, and finally the settings page.
+
+In most environments where both Angular and AngularJS are used to render the application, both frameworks are loaded in the initial bundle being sent to the client. This not only results in a larger bundle size, but also means if a user only stays on Angular-rendered pages, the AngularJS framework and application is still loaded, and running, even if it is never accessed, potentially impacting ov ...
```
This paragraph seems to be mixing too different aspects: (a) loading the AngularJS parts lazily and (b) bootstrapping the AngularJS app lazily.
They are not the same thing and impact the app differently.

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -838,6 +838,59 @@ After this, the service is injectable anywhere in AngularJS code:
+
+### Create a service to lazy load AngularJS
+
+As of Angular version 8, lazy loading code can be accomplished simply by using the dynamic import syntax `import('..')`. In your application, you create a new service that uses dynamic imports to lazy load AngularJS.
```
_import('..')_ --> _import('...')_

gkalpak commented on aio/content/examples/upgrade-lazy-load-ajs/src/app/lazy-loader.service.ts:
```diff
@@ -0,0 +1,19 @@
+export class LazyLoaderService {
+  bootstrapped = false;
+
+  load(el?: HTMLElement): void {
```
Why make `el` optional, when it is required for `app.bootstrap()`?
Also, why doesn't TS complain about that? 😕

brandonroberts replied:
Good catch. Not sure why it doesn't complain there

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -838,6 +838,59 @@ After this, the service is injectable anywhere in AngularJS code:
+<code-example path="upgrade-lazy-load-ajs/src/app/lazy-loader.service.ts" header="src/app/lazy-loader.service.ts">
+</code-example>
+
+The service uses the `import()` method to load your bundled AngularJS application lazily. This decreases the initial bundle size of your application as you're not loading code your user doesn't need yet. You also need to provide a way to _bootstrap_ the application manually after it has been loaded. AngularJS provides a way to manually bootstrap an application using the `angular.bootstrap()` meth ...
```
It would be useful to make `angular.bootstrap()` a link to the AngularJS docs (https://docs.angularjs.org/api/ng/function/angular.bootstrap).

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -838,6 +838,59 @@ After this, the service is injectable anywhere in AngularJS code:
+
+<div class="alert is-important">
+
+**Note:** After AngularJS is loaded and bootstrapped, listeners such as those wired up in your route configuration will continue to listen for route changes. To ensure listeners are shut down when AngularJS isn’t being displayed, configure an `otherwise` option with the `$routeProvider` that renders an empty template. This assumes all other routes will be handled by Angular.
```
> isn’t

Fancy quote 🙈 

##
Also, a [link to `$routeProvider][1] would be useful.

[1]: https://docs.angularjs.org/api/ngRoute/provider/$routeProvider

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -838,6 +838,59 @@ After this, the service is injectable anywhere in AngularJS code:
+
+### Configure a custom route matcher for AngularJS routes
+
+To configure the Angular router, you must define a route for AngularJS URLs. To match those URLs, you add a route configuration that uses the `matcher` property. The `matcher` allows you to use custom pattern matching for URL paths. The Angular router tries to match on more specific routes such as static and variable routes first. When it doesn’t find a match, it then looks at custom matchers def ...
```
> doesn’t

Fancy quote 🙈

gkalpak reviewed (dismissed):
This is awesome :tada:
Left a few comments, plus: 

- Please, add the new files (e.g. example) to `.github/CODEOWNERS`.
- `UpgradeModule` is not used here (afaict). Although this might be a viable upgrade strategy, this is essentially not an `ngUpgrade` app.
- Where are the tests? 😱 😱 😱 

I am borderline inclined to say that this should be its own guide.
We are also talking about _"using the cli for bundling"_ like ain't nothin', but I don't think we show the users how to use the cli for a hybrid app (and it _is_ a big deal). (Just to be clear: We totally should document using the cli for hybrid apps, we just haven't done it yet.)

brandonroberts commented:
> This is awesome 🎉
> Left a few comments, plus:
> 
> * Please, add the new files (e.g. example) to `.github/CODEOWNERS`.
> * `UpgradeModule` is not used here (afaict). Although this might be a viable upgrade strategy, this is essentially not an `ngUpgrade` app.
> * Where are the tests? 😱 😱 😱
> 
> I am borderline inclined to say that this should be its own guide.
> We are also talking about _"using the cli for bundling"_ like ain't nothin', but I don't think we show the users how to use the cli for a hybrid app (and it _is_ a big deal). (Just to be clear: We totally should document using the cli for hybrid apps, we just haven't done it yet.)

Thanks @gkalpak! 

- I agree, the bundling part is very hand-wavy, and that should be a follow-up as a bigger part of the rethink of the upgrade guide.
- Added example to `.github/CODEOWNERS`
- Removed UpgradeModule
- Using dynamic imports requires `esnext` in the `tsconfig.json`. I was going to do the tests in a follow-up PR after #30385 lands so no extra changes are needed

jbogarthyde commented on aio/content/guide/upgrade.md:
```diff
@@ -838,6 +838,59 @@ After this, the service is injectable anywhere in AngularJS code:
+To configure the Angular router, you must define a route for AngularJS URLs. To match those URLs, you add a route configuration that uses the `matcher` property. The `matcher` allows you to use custom pattern matching for URL paths. The Angular router tries to match on more specific routes such as static and variable routes first. When it doesn’t find a match, it then looks at custom matchers def ...
+
+Define a custom matcher function for AngularJS routes:
+
```
The following example defines a custom matcher function for AngularJS routes.

jbogarthyde commented on aio/content/guide/upgrade.md:
```diff
@@ -838,6 +838,59 @@ After this, the service is injectable anywhere in AngularJS code:
+</code-example>
+
+Then add a route object to your routing configuration using the `matcher` property and custom matcher, and the `component` property with `AngularJSComponent`.
+
```
The following code adds a route object ...

jbogarthyde reviewed (commented):
copy edit/slight re-org

brandonroberts committed:
fixup! docs: add section to upgrade guide on lazy loading AngularJS

brandonroberts committed:
fixup! docs: add section to upgrade guide on lazy loading AngularJS

brandonroberts committed:
fixup! docs: add section to upgrade guide on lazy loading AngularJS

brandonroberts committed:
fixup! docs: add section to upgrade guide on lazy loading AngularJS

mary-poppins commented:
You can preview a684746 at https://pr30541-a684746.ngbuilds.io/.

mary-poppins commented:
You can preview 8790c5c at https://pr30541-8790c5c.ngbuilds.io/.

mary-poppins commented:
You can preview 169dc4d at https://pr30541-169dc4d.ngbuilds.io/.

mary-poppins commented:
You can preview 15b8503 at https://pr30541-15b8503.ngbuilds.io/.

brandonroberts committed:
fixup! docs: add section to upgrade guide on lazy loading AngularJS

mary-poppins commented:
You can preview 6c6bac3 at https://pr30541-6c6bac3.ngbuilds.io/.

jasonaden reviewed (dismissed):
Agreed, this is great and I think it will really help people in this situation.

Reading it from a user's perspective, I think there's a couple things missing that could be solved by an introduction section. The problem description is there, but I think we should beef it up a little bit to make it more clear exactly what the problem is we're solving here.

The other thing I think would be really useful is an overview of the process we're going to use. There are a few distinct parts (not necessarily in this order):

* Setting up your AngularJS application to be lazy loaded (for the bootstrap callback)
* Create the lazy loading service
* Create a component to render your lazy loaded application
* Define a matcher function to determine if a URL is handled by AngularJS or not
* Configure routing

I think adding some sort of intro discussing the overall process would be really helpful for context while reading.

samjulien commented on aio/content/guide/upgrade.md:
```diff
@@ -838,6 +838,63 @@ After this, the service is injectable anywhere in AngularJS code:
+
+In most environments where both Angular and AngularJS are used to render the application, both frameworks are loaded in the initial bundle being sent to the client. This results in both increased bundle size and possible reduced performance.
+
+Overall application performance is affected in cases where the user stays on Angular-rendered pages, because the AngularJS framework and application is still loaded, and running, even if it is never accessed.
```
```suggestion
Overall application performance is affected in cases where the user stays on Angular-rendered pages because the AngularJS framework and application are still loaded and running, even if they are never accessed.
```

samjulien commented on aio/content/guide/upgrade.md:
```diff
@@ -838,6 +838,63 @@ After this, the service is injectable anywhere in AngularJS code:
+
+Overall application performance is affected in cases where the user stays on Angular-rendered pages, because the AngularJS framework and application is still loaded, and running, even if it is never accessed.
+
+To address the performance problem, you can use  [lazy loading](guide/glossary#lazy-loading) to only load, bootstrap, and render the AngularJS application when needed. This has the added benefit of reducing the initial bundle size.
```
```suggestion
To address the performance problem, you can use  [lazy loading](guide/glossary#lazy-loading) to load, bootstrap, and render the AngularJS application only when needed. This has the added benefit of reducing the initial bundle size.
```

samjulien commented on aio/content/guide/upgrade.md:
```diff
@@ -838,6 +838,63 @@ After this, the service is injectable anywhere in AngularJS code:
+
+To address the performance problem, you can use  [lazy loading](guide/glossary#lazy-loading) to only load, bootstrap, and render the AngularJS application when needed. This has the added benefit of reducing the initial bundle size.
+
+The steps below steps below show you how to create a service that lazy loads and boostraps AngularJS, create a component for AngularJS content, and how to configure the Angular Router to load AngularJS routes only on specific URLs.
```
```suggestion
The steps below show you how to create a service that lazy loads and bootstraps AngularJS, create a component for AngularJS content, and configure the Angular Router to load AngularJS routes only on specific URLs.
```

samjulien commented on aio/content/guide/upgrade.md:
```diff
@@ -838,6 +838,63 @@ After this, the service is injectable anywhere in AngularJS code:
+
+### Configure a custom route matcher for AngularJS routes
+
+To configure the Angular router, you must define a route for AngularJS URLs. To match those URLs, you add a route configuration that uses the `matcher` property. The `matcher` allows you to use custom pattern matching for URL paths. The Angular router tries to match on more specific routes such as static and variable routes first. When it doesn't find a match, it then looks at custom matchers def ...
```
Should it always be Angular Router (capitalized)?

```suggestion
To configure the Angular Router, you must define a route for AngularJS URLs. To match those URLs, you add a route configuration that uses the `matcher` property. The `matcher` allows you to use custom pattern matching for URL paths. The Angular Router tries to match on more specific routes such as static and variable routes first. When it doesn't find a match, it then looks at custom matchers defined in your route configuration. If the custom matchers don't match a route, it then goes to catch-all routes, such as a 404 page.
```

jbogarthyde replied:
It should only be capitalized (and put it code font) when referring to the API element. When used descriptively like this, it should remain lowercase.

samjulien reviewed (commented):
Just a few typos and corrections. I also agree with George on the need to document bundling/setting up the CLI with ngUpgrade, which I hope to be involved with.

brandonroberts committed:
fixup! docs: add section to upgrade guide on lazy loading AngularJS

mary-poppins commented:
You can preview 61c1311 at https://pr30541-61c1311.ngbuilds.io/.

mary-poppins commented:
You can preview 03a9670 at https://pr30541-03a9670.ngbuilds.io/.

mary-poppins commented:
You can preview 269eec2 at https://pr30541-269eec2.ngbuilds.io/.

brandonroberts committed:
fixup! docs: add section to upgrade guide on lazy loading AngularJS

mary-poppins commented:
You can preview 80a243c at https://pr30541-80a243c.ngbuilds.io/.

IgorMinar reviewed (approved):
this looks good to me! thanks @brandonroberts

jbogarthyde commented on aio/content/guide/upgrade.md:
```diff
@@ -838,6 +838,63 @@ After this, the service is injectable anywhere in AngularJS code:
+## Lazy Loading AngularJS
+
+When building applications, you want to ensure that only the required resources are loaded when necessary. Whether that be loading of assets or code, making sure everything that can be deferred until needed keeps your application running efficiently. This is especially true when running different frameworks in the same application. When migrating large applications from AngularJS to Angular using ...
+
```
[Lazy loading](guide/glossary#lazy-loading) is a technique that defers the loading of required assets and code resources until they are actually used. This reduces startup time and increases efficiency, especially when running different frameworks in the same application. 

When migrating...

jbogarthyde commented on aio/content/guide/upgrade.md:
```diff
@@ -838,6 +838,63 @@ After this, the service is injectable anywhere in AngularJS code:
+Overall application performance is affected in cases where the user stays on Angular-rendered pages because the AngularJS framework and application are still loaded and running, even if they are never accessed. 
+
+You can take steps to mitigate these bundle size and performance issues. By isolating your AngularJS app to a separate bundle, you can use [lazy loading](guide/glossary#lazy-loading) to load, bootstrap, and render the AngularJS application only when needed. This reduces your initial bundle size, defers any potential impact from loading both frameworks until absolutlely necessary, and keeps your a ...
+
```
You can take steps to mitigate both bundle size and performance issues. By isolating your AngularJS app to a separate bundle, you can take advantage of lazy loading to load, bootstrap, and render the AngularJS application only when needed. This reduces your initial bundle size, defers any potential impact from loading both frameworks until absolutely necessary, and keeps your application running as efficiently as possible.

jbogarthyde commented on aio/content/guide/upgrade.md:
```diff
@@ -838,6 +838,63 @@ After this, the service is injectable anywhere in AngularJS code:
+You can take steps to mitigate these bundle size and performance issues. By isolating your AngularJS app to a separate bundle, you can use [lazy loading](guide/glossary#lazy-loading) to load, bootstrap, and render the AngularJS application only when needed. This reduces your initial bundle size, defers any potential impact from loading both frameworks until absolutlely necessary, and keeps your a ...
+
+The steps below show you how to setup a callback function for your AngularJS bundle, create a service that lazy loads and bootstraps your AngularJS app, create a routable component for AngularJS content, create a custom matcher function for AngularJS-specific URLs, and configure the Angular Router with the custom matcher for AngularJS routes.
+
```
Make this a bullet list: 

The step below show you how to do the following:

* Set up a callback function for your AngularJS bundle.
* Create a service that lazy loads and bootstraps your AngularJS app.
* Create a routable component for AngularJS content.
* Create a custom `matcher` function for AngularJS-specific URLs and configure the Angular `Router` with the custom matcher for AngularJS routes.

jbogarthyde reviewed (changes requested):
Copy edit/organization

brandonroberts committed:
fixup! docs: add section to upgrade guide on lazy loading AngularJS

mary-poppins commented:
You can preview 8a73bfd at https://pr30541-8a73bfd.ngbuilds.io/.

jbogarthyde reviewed (approved):
LGTM
